src/validation/ValidationContext.js

```diff
@@ -42,11 +42,44 @@ type VariableUsage = {|
  * allowing access to commonly useful contextual information from within a
  * validation rule.
  */
-export default class ValidationContext {
-  _schema: GraphQLSchema;
+export class BaseValidationContext {
   _ast: DocumentNode;
-  _typeInfo: TypeInfo;
   _errors: Array<GraphQLError>;
+
+  constructor(ast: DocumentNode): void {
+    this._ast = ast;
+    this._errors = [];
+  }
+
+  reportError(error: GraphQLError): void {
+    this._errors.push(error);
+  }
+
+  getErrors(): $ReadOnlyArray<GraphQLError> {
+    return this._errors;
+  }
+
+  getDocument(): DocumentNode {
+    return this._ast;
+  }
+}
+
+export class SDLValidationContext extends BaseValidationContext {
+  _schema: ?GraphQLSchema;
+
+  constructor(ast: DocumentNode, schema?: ?GraphQLSchema): void {
+    super(ast);
+    this._schema = schema;
```

Comment:
I don't think we should even allow access to a `GraphQLSchema` here: instead, if you pass in an existing schema, we can convert that schema to AST nodes (a very rough way to do this would be to `parse(printSchema(schema))`).

Ideally, the current `GraphQLSchema` would only be a runtime thing: you need it to validate queries you're trying to persist, but you don't need it to validate or transform itself.

I like the idea that there's a split between "validation that needs a schema" and "validation that does not", and "validation that does not need a schema" is a superset of "SDL validation".

Reply:
@mjmahone Problem is that conversion from SDL to `GraphQLSchema` is lossless, but not vice versa. For example, you lose all directives which prevent from enforcing the rule that forbids directive duplication in extends.

Also after `parse(printSchema(schema)))` you loose ability to print errors with original locations and formating of nodes.

And there are others many potential scenarious where we loose something by such converion especially if schema that we trying to extend was build from `GraphQL*Types` not from SDL. For example, developer can create custom class derived `GraphQLObjectType` with some additional data and require this additional data inside his custom SDL rules.

And it's not a technical problem, since `GraphQLSchema` is intended to be most complete represenation of schema.

> and "validation that does not need a schema" is a superset of "SDL validation".

"validation that does not need a schema" => ASTValidationContext
And this context suffisient for all rules that scoped to particular defintion.

If rule should check conflicts/dependencies beetween definitions it should handle case when you extending existing schema and use `SDLValidationContext`.

Reply:
We should fix it so it's lossless, but that would be a different PR, and we could probably target 14.1 for that.

> Also after parse(printSchema(schema))) you loose ability to print errors with original locations and formating of nodes.

I think there's a way around this, where we'd do `buildSchemaDocuments(schema): Array<DocumentNode>`, where each `DocumentNode` would contain all definitions from a specific source. That's obviously more complex of a change than what we want to do here, but I think would be worth having. For `GraphQL*Types` we'd have to have some form of "default" or "non-Source" document. But for truly custom classes, you're right we can't make lossless conversions.

> If rule should check conflicts/dependencies beetween definitions it should handle case when you extending existing schema and use SDLValidationContext.

Makes sense, so let's push this forward as-is. But I'd like to switch our mental model to being AST-validation first, and then only using the schema when it's truly necessary. You've mostly accomplished this, but I envision a bunch of people creating validation rules that silently fail if no schema is present (because they assume they always have a schema present). That would be an unfortunate future.